File: app/internal/publisher/template/dotnet/ApiClient.cs
using System;
using System.Collections.Generic;
using System.IO;
using System.Net.Http;
using System.Text;
using System.Text.Json;
using System.Threading;
using System.Threading.Tasks;

namespace WebsmithSDK;

public class ApiException : Exception
{
    public int StatusCode { get; }
    public string? ResponseBody { get; }

    public ApiException(string message, int statusCode = 0, string? responseBody = null)
        : base(message)
    {
        StatusCode = statusCode;
        ResponseBody = responseBody;
    }
}

public class ApiClient
{
    private static readonly HashSet<int> RetryableStatuses = new() { 500, 502, 503, 504 };

    private readonly string _apiKey;
    private readonly string _apiSecret;
    private readonly string _baseUrl;
    private readonly string _apiVersion;
    private readonly string _productId;
    private readonly int _retryCount;
    private readonly int _timeoutMs;
    private readonly HttpClient _httpClient;
    private HardwareDetector _hardware;
    private CacheManager _cache;

    public ApiClient() : this("config/api-config.json") { }

    public ApiClient(string configPath)
    {
        var config = LoadConfigFile(configPath);
        var api = config.GetProperty("api");
        var product = config.GetProperty("product");
        _apiKey = GetJsonString(api, "public_key");
        _apiSecret = GetJsonString(api, "secret");
        var envUrl = Environment.GetEnvironmentVariable("WEBSMITH_API_URL");
        _baseUrl = (envUrl ?? GetJsonString(api, "url")).TrimEnd('/');
        _apiVersion = GetJsonString(api, "version");
        if (string.IsNullOrEmpty(_apiVersion)) _apiVersion = "v1";
        _productId = GetJsonString(product, "id");
        _retryCount = GetJsonInt(api, "retry_count", 3);
        _timeoutMs = GetJsonInt(api, "timeout", 30000);
        _httpClient = new HttpClient { Timeout = TimeSpan.FromMilliseconds(_timeoutMs) };
        _hardware = new HardwareDetector();
        _cache = new CacheManager(config);
    }

    public ApiClient(JsonDocument config, HardwareDetector hardware, CacheManager cache)
    {
        var api = config.RootElement.GetProperty("api");
        _apiKey = GetJsonString(api, "public_key");
        _apiSecret = GetJsonString(api, "secret");
        var envUrl = Environment.GetEnvironmentVariable("WEBSMITH_API_URL");
        _baseUrl = (envUrl ?? GetJsonString(api, "url")).TrimEnd('/');
        _apiVersion = GetJsonString(api, "version");
        if (string.IsNullOrEmpty(_apiVersion)) _apiVersion = "v1";
        _productId = GetJsonString(config.RootElement.GetProperty("product"), "id");
        _retryCount = GetJsonInt(api, "retry_count", 3);
        _timeoutMs = GetJsonInt(api, "timeout", 30000);
        _httpClient = new HttpClient { Timeout = TimeSpan.FromMilliseconds(_timeoutMs) };
        _hardware = hardware;
        _cache = cache;
    }

    public HardwareDetector GetHardware() => _hardware;
    public CacheManager GetCache() => _cache;

    private string GetHardwareId() => _hardware.GetFingerprint();

    private static JsonDocument LoadConfigFile(string path)
    {
        if (File.Exists(path))
        {
            var text = File.ReadAllText(path);
            return JsonDocument.Parse(text);
        }
        var fallback = $$"""
        {
            "api": { "public_key": "", "secret": "", "url": "", "version": "v1", "retry_count": 3, "timeout": 30000 },
            "product": { "id": "" }
        }
        """;
        return JsonDocument.Parse(fallback);
    }

    private static string GetJsonString(JsonElement obj, string key)
    {
        if (obj.ValueKind == JsonValueKind.Object && obj.TryGetProperty(key, out var prop))
        {
            return prop.GetString() ?? "";
        }
        return "";
    }

    private static int GetJsonInt(JsonElement obj, string key, int def)
    {
        if (obj.ValueKind == JsonValueKind.Object && obj.TryGetProperty(key, out var prop))
        {
            try { return prop.GetInt32(); } catch { }
        }
        return def;
    }

    private async Task<JsonDocument> RequestAsync(string endpoint, Dictionary<string, object> data)
    {
        var url = $"{_baseUrl}/api/{_apiVersion}/{endpoint}";
        var apiPath = $"/api/{_apiVersion}/{endpoint}";
        var maxRetries = _retryCount;

        if (!string.IsNullOrEmpty(_productId) && !data.ContainsKey("product_id"))
        {
            data["product_id"] = _productId;
        }

        for (int attempt = 0; attempt <= maxRetries; attempt++)
        {
            try
            {
                var json = JsonSerializer.Serialize(data);
                var timestamp = CryptoUtils.GenerateTimestamp();
                var nonce = CryptoUtils.GenerateNonce();
                var signature = CryptoUtils.SignRequest(json, _apiSecret, timestamp, nonce, "POST", apiPath, "");

                var content = new StringContent(json, Encoding.UTF8, "application/json");
                content.Headers.Add("X-API-Key", _apiKey);
                content.Headers.Add("X-Timestamp", timestamp);
                content.Headers.Add("X-Nonce", nonce);
                content.Headers.Add("X-Signature", signature);

                var response = await _httpClient.PostAsync(url, content);
                var responseBody = await response.Content.ReadAsStringAsync();

                JsonDocument resultDoc;
                try
                {
                    resultDoc = JsonDocument.Parse(responseBody);
                }
                catch
                {
                    var fallback = JsonSerializer.Serialize(new { message = responseBody });
                    resultDoc = JsonDocument.Parse(fallback);
                }

                if ((int)response.StatusCode >= 200 && (int)response.StatusCode < 300)
                {
                    return resultDoc;
                }

                if ((int)response.StatusCode == 429)
                {
                    if (attempt < maxRetries)
                    {
                        await Task.Delay(5000);
                        continue;
                    }
                    throw new ApiException("Rate limit exceeded", 429, responseBody);
                }

                if (RetryableStatuses.Contains((int)response.StatusCode))
                {
                    if (attempt < maxRetries)
                    {
                        await Task.Delay((attempt + 1) * 2000);
                        continue;
                    }
                }

                var root = resultDoc.RootElement;
                var msg = root.TryGetProperty("message", out var m) ? m.GetString()
                    : root.TryGetProperty("error", out var e) ? e.GetString()
                    : $"HTTP {(int)response.StatusCode}";
                throw new ApiException(msg ?? "Request failed", (int)response.StatusCode, responseBody);
            }
            catch (ApiException) { throw; }
            catch (TaskCanceledException)
            {
                if (attempt < maxRetries)
                {
                    await Task.Delay((attempt + 1) * 2000);
                    continue;
                }
                throw new ApiException("Request timeout", 504);
            }
            catch (HttpRequestException ex)
            {
                if (attempt < maxRetries)
                {
                    await Task.Delay((attempt + 1) * 2000);
                    continue;
                }
                throw new ApiException($"Connection error: {ex.Message}", 503);
            }
            catch (Exception ex)
            {
                if (attempt < maxRetries)
                {
                    await Task.Delay((attempt + 1) * 2000);
                    continue;
                }
                throw new ApiException($"Request failed: {ex.Message}", 500);
            }
        }
        throw new ApiException($"Failed after {maxRetries} retries", 500);
    }

    public async Task<JsonDocument> ValidateLicense(string licenseKey, string hardwareId)
    {
        var data = new Dictionary<string, object>
        {
            ["action"] = "validate",
            ["license_key"] = licenseKey,
            ["hardware_id"] = hardwareId
        };
        if (_cache != null && _cache.IsValid())
        {
            var cached = _cache.GetLicenseStatus();
            if (cached != null)
            {
                return JsonDocument.Parse(cached);
            }
        }
        var response = await RequestAsync("license", data);
        if (_cache != null && response.RootElement.TryGetProperty("success", out var s) && s.GetBoolean()
            && response.RootElement.TryGetProperty("data", out var d) && d.TryGetProperty("valid", out var v) && v.GetBoolean())
        {
            _cache.SetLicenseStatus(response.RootElement.GetRawText());
        }
        return response;
    }

    public async Task<JsonDocument> ActivateLicense(string licenseKey, string hardwareId)
    {
        var data = new Dictionary<string, object>
        {
            ["action"] = "activate",
            ["license_key"] = licenseKey,
            ["hardware_id"] = hardwareId
        };
        var response = await RequestAsync("license", data);
        _cache?.InvalidateLicenseStatus();
        return response;
    }

    public async Task<JsonDocument> DeactivateLicense(string licenseKey, string hardwareId)
    {
        var data = new Dictionary<string, object>
        {
            ["action"] = "deactivate",
            ["license_key"] = licenseKey,
            ["hardware_id"] = hardwareId
        };
        var response = await RequestAsync("license", data);
        _cache?.InvalidateLicenseStatus();
        return response;
    }

    public async Task<JsonDocument> RenewLicense(string licenseKey, int? extraDays = null)
    {
        var data = new Dictionary<string, object>
        {
            ["action"] = "renew",
            ["license_key"] = licenseKey
        };
        if (extraDays.HasValue)
        {
            data["extra_days"] = extraDays.Value;
        }
        var response = await RequestAsync("license", data);
        _cache?.InvalidateLicenseStatus();
        return response;
    }

    public async Task<JsonDocument> StartTrial(string email, string customerName, string hardwareId, Dictionary<string, object>? customerData = null)
    {
        var data = new Dictionary<string, object>
        {
            ["action"] = "start",
            ["customer_email"] = email,
            ["customer_name"] = customerName ?? "",
            ["hardware_id"] = hardwareId
        };
        if (customerData != null)
        {
            foreach (var kv in customerData)
                data[kv.Key] = kv.Value;
        }
        return await RequestAsync("trial", data);
    }

    public async Task<JsonDocument> GetTrialStatus(string hardwareId)
    {
        var data = new Dictionary<string, object>
        {
            ["action"] = "status",
            ["hardware_id"] = hardwareId
        };
        return await RequestAsync("trial", data);
    }

    public async Task<JsonDocument> ConvertTrial(string hardwareId, string? plan, string customerName, string customerEmail)
    {
        var data = new Dictionary<string, object>
        {
            ["action"] = "convert",
            ["hardware_id"] = hardwareId
        };
        if (!string.IsNullOrEmpty(plan)) data["plan"] = plan;
        if (!string.IsNullOrEmpty(customerName)) data["customer_name"] = customerName;
        if (!string.IsNullOrEmpty(customerEmail)) data["customer_email"] = customerEmail;
        var response = await RequestAsync("trial", data);
        _cache?.InvalidateLicenseStatus();
        return response;
    }

    public async Task<JsonDocument> ReplaceDevice(string licenseKey, string oldHardwareId, string newHardwareId)
    {
        var data = new Dictionary<string, object>
        {
            ["action"] = "replace",
            ["license_key"] = licenseKey,
            ["old_hardware_id"] = oldHardwareId,
            ["new_hardware_id"] = newHardwareId
        };
        var response = await RequestAsync("device", data);
        _cache?.InvalidateLicenseStatus();
        return response;
    }

    public async Task<JsonDocument> BindDevice(string licenseKey, string hardwareId, string? deviceName = null)
    {
        var data = new Dictionary<string, object>
        {
            ["action"] = "bind",
            ["license_key"] = licenseKey,
            ["hardware_id"] = hardwareId
        };
        if (!string.IsNullOrEmpty(deviceName)) data["device_name"] = deviceName;
        return await RequestAsync("device", data);
    }

    public async Task<JsonDocument> GetProducts()
    {
        var data = new Dictionary<string, object>
        {
            ["action"] = "list"
        };
        if (!string.IsNullOrEmpty(_productId))
            data["product_id"] = _productId;
        try
        {
            return await RequestAsync("store/products", data);
        }
        catch
        {
            return JsonDocument.Parse("{\"success\":false,\"products\":[]}");
        }
    }

    public async Task<JsonDocument> UpdateCustomer(string name, string email, string phone, string hardwareId)
    {
        var data = new Dictionary<string, object>
        {
            ["action"] = "update",
            ["name"] = name,
            ["email"] = email,
            ["mobile"] = phone,
            ["hardware_id"] = hardwareId
        };
        try
        {
            return await RequestAsync("customer/register", data);
        }
        catch (Exception ex)
        {
            return JsonDocument.Parse($"{{\"success\":false,\"error\":\"{ex.Message}\"}}");
        }
    }
}

File: app/internal/publisher/template/java/ApiClient.java
package com.websmith.sdk;

import com.google.gson.Gson;
import com.google.gson.JsonObject;
import com.google.gson.JsonParser;
import java.io.File;
import java.io.FileReader;
import java.io.InputStreamReader;
import java.net.URI;
import java.net.http.HttpClient;
import java.net.http.HttpRequest;
import java.net.http.HttpResponse;
import java.nio.charset.StandardCharsets;
import java.time.Duration;
import java.util.HashSet;
import java.util.Set;

public class ApiClient {

    private static final Set<Integer> RETRYABLE_STATUSES = new HashSet<>();
    static {
        RETRYABLE_STATUSES.add(500);
        RETRYABLE_STATUSES.add(502);
        RETRYABLE_STATUSES.add(503);
        RETRYABLE_STATUSES.add(504);
    }

    private final String apiKey;
    private final String apiSecret;
    private final String baseUrl;
    private final String apiVersion;
    private final String productId;
    private final int retryCount;
    private final int timeoutMs;
    private final HttpClient httpClient;
    private final Gson gson;
    private HardwareDetector hardware;
    private CacheManager cache;

    public ApiClient() {
        this("config/api-config.json");
    }

    public ApiClient(String configPath) {
        JsonObject config = loadConfigFile(configPath);
        JsonObject api = config.has("api") ? config.getAsJsonObject("api") : new JsonObject();
        JsonObject product = config.has("product") ? config.getAsJsonObject("product") : new JsonObject();
        this.apiKey = getJsonString(api, "public_key");
        this.apiSecret = getJsonString(api, "secret");
        String envUrl = System.getenv("WEBSMITH_API_URL");
        this.baseUrl = (envUrl != null ? envUrl : getJsonString(api, "url")).replaceAll("/+$", "");
        this.apiVersion = getJsonString(api, "version");
        if (this.apiVersion.isEmpty()) this.apiVersion = "v1";
        this.productId = getJsonString(product, "id");
        this.retryCount = getJsonInt(api, "retry_count", 3);
        this.timeoutMs = getJsonInt(api, "timeout", 30000);
        this.httpClient = HttpClient.newBuilder()
            .connectTimeout(Duration.ofMillis(this.timeoutMs))
            .build();
        this.gson = new Gson();
        this.hardware = new HardwareDetector();
        this.cache = new CacheManager(config);
    }

    public ApiClient(JsonObject config, HardwareDetector hardware, CacheManager cache) {
        JsonObject api = config.has("api") ? config.getAsJsonObject("api") : new JsonObject();
        JsonObject product = config.has("product") ? config.getAsJsonObject("product") : new JsonObject();
        this.apiKey = getJsonString(api, "public_key");
        this.apiSecret = getJsonString(api, "secret");
        String envUrl = System.getenv("WEBSMITH_API_URL");
        this.baseUrl = (envUrl != null ? envUrl : getJsonString(api, "url")).replaceAll("/+$", "");
        this.apiVersion = getJsonString(api, "version");
        if (this.apiVersion.isEmpty()) this.apiVersion = "v1";
        this.productId = getJsonString(product, "id");
        this.retryCount = getJsonInt(api, "retry_count", 3);
        this.timeoutMs = getJsonInt(api, "timeout", 30000);
        this.httpClient = HttpClient.newBuilder()
            .connectTimeout(Duration.ofMillis(this.timeoutMs))
            .build();
        this.gson = new Gson();
        this.hardware = hardware;
        this.cache = cache;
    }

    private static JsonObject loadConfigFile(String path) {
        try {
            File f = new File(path);
            if (f.exists()) {
                try (FileReader reader = new FileReader(f)) {
                    return JsonParser.parseReader(reader).getAsJsonObject();
                }
            }
        } catch (Exception ignored) {}
        try (InputStreamReader reader = new InputStreamReader(
                ApiClient.class.getClassLoader().getResourceAsStream(path))) {
            if (reader != null) {
                return JsonParser.parseReader(reader).getAsJsonObject();
            }
        } catch (Exception ignored) {}
        JsonObject fallback = new JsonObject();
        JsonObject api = new JsonObject();
        api.addProperty("public_key", "");
        api.addProperty("secret", "");
        api.addProperty("url", "");
        api.addProperty("version", "v1");
        api.addProperty("retry_count", 3);
        api.addProperty("timeout", 30000);
        fallback.add("api", api);
        fallback.add("product", new JsonObject());
        return fallback;
    }

    private static String getJsonString(JsonObject obj, String key) {
        if (obj != null && obj.has(key) && !obj.get(key).isJsonNull()) {
            return obj.get(key).getAsString();
        }
        return "";
    }

    private static int getJsonInt(JsonObject obj, String key, int def) {
        if (obj != null && obj.has(key) && !obj.get(key).isJsonNull()) {
            try { return obj.get(key).getAsInt(); } catch (Exception ignored) {}
        }
        return def;
    }

    public static class ApiException extends Exception {
        private final int statusCode;
        private final JsonObject responseData;

        public ApiException(int statusCode, String message) {
            this(statusCode, message, null);
        }

        public ApiException(int statusCode, String message, JsonObject data) {
            super(message);
            this.statusCode = statusCode;
            this.responseData = data;
        }

        public int getStatusCode() { return statusCode; }
        public JsonObject getResponseData() { return responseData; }
    }

    private String getHardwareId() {
        return hardware.getFingerprint();
    }

    public HardwareDetector getHardware() {
        return hardware;
    }

    public CacheManager getCache() {
        return cache;
    }

    private JsonObject signRequestHeaders(JsonObject payload, String method, String path, String query)
            throws Exception {
        String timestamp = CryptoUtils.generateTimestamp();
        String nonce = CryptoUtils.generateNonce();
        String signature = CryptoUtils.signRequest(payload, apiSecret, timestamp, nonce, method, path, query);
        JsonObject headers = new JsonObject();
        headers.addProperty("x-api-key", apiKey);
        headers.addProperty("x-timestamp", timestamp);
        headers.addProperty("x-nonce", nonce);
        headers.addProperty("x-signature", signature);
        return headers;
    }

    private JsonObject request(String endpoint, JsonObject data) throws ApiException {
        String url = baseUrl + "/api/" + apiVersion + "/" + endpoint;
        String apiPath = "/api/" + apiVersion + "/" + endpoint;
        int maxRetries = retryCount;

        if (productId != null && !productId.isEmpty() && !data.has("product_id")) {
            data.addProperty("product_id", productId);
        }

        for (int attempt = 0; attempt <= maxRetries; attempt++) {
            try {
                String timestamp = CryptoUtils.generateTimestamp();
                String nonce = CryptoUtils.generateNonce();
                String signature = CryptoUtils.signRequest(data, apiSecret, timestamp, nonce, "POST", apiPath, "");

                String bodyJson = gson.toJson(data);
                HttpRequest request = HttpRequest.newBuilder()
                    .uri(URI.create(url))
                    .header("Content-Type", "application/json")
                    .header("X-API-Key", apiKey)
                    .header("X-Timestamp", timestamp)
                    .header("X-Nonce", nonce)
                    .header("X-Signature", signature)
                    .method("POST", HttpRequest.BodyPublishers.ofString(bodyJson, StandardCharsets.UTF_8))
                    .build();

                HttpResponse<String> response = httpClient.send(request, HttpResponse.BodyHandlers.ofString());
                int status = response.statusCode();
                String responseBody = response.body();

                JsonObject result;
                try {
                    result = JsonParser.parseString(responseBody).getAsJsonObject();
                } catch (Exception e) {
                    result = new JsonObject();
                    result.addProperty("message", responseBody);
                }

                if (status >= 200 && status < 300) {
                    return result;
                }

                if (status == 429) {
                    if (attempt < maxRetries) {
                        int retryAfter = 5;
                        if (result.has("retry_after")) {
                            try { retryAfter = result.get("retry_after").getAsInt(); } catch (Exception ignored) {}
                        }
                        Thread.sleep(retryAfter * 1000L);
                        continue;
                    }
                    throw new ApiException(status, "Rate limit exceeded", result);
                }

                if (RETRYABLE_STATUSES.contains(status)) {
                    if (attempt < maxRetries) {
                        Thread.sleep((long) ((attempt + 1) * 2 * 1000L));
                        continue;
                    }
                }

                String msg = result.has("message") ? result.get("message").getAsString()
                    : result.has("error") ? result.get("error").getAsString()
                    : "HTTP " + status;
                throw new ApiException(status, msg, result);

            } catch (ApiException e) {
                throw e;
            } catch (InterruptedException e) {
                Thread.currentThread().interrupt();
                throw new ApiException(500, "Request interrupted");
            } catch (Exception e) {
                if (attempt < maxRetries) {
                    try { Thread.sleep((long) ((attempt + 1) * 2 * 1000L)); } catch (InterruptedException ie) {
                        Thread.currentThread().interrupt();
                        throw new ApiException(500, "Request interrupted");
                    }
                    continue;
                }
                throw new ApiException(500, "Request failed after " + maxRetries + " retries: " + e.getMessage());
            }
        }
        throw new ApiException(500, "Request failed after " + maxRetries + " retries");
    }

    public JsonObject validateLicense(String licenseKey, String hardwareId) throws ApiException {
        JsonObject data = new JsonObject();
        data.addProperty("action", "validate");
        data.addProperty("license_key", licenseKey);
        data.addProperty("hardware_id", hardwareId);
        if (cache != null && cache.isValid()) {
            Object cached = cache.get("license_status");
            if (cached instanceof String) {
                try {
                    return JsonParser.parseString((String) cached).getAsJsonObject();
                } catch (Exception ignored) {}
            }
        }
        JsonObject response = request("license", data);
        if (cache != null && response.has("success") && response.get("success").getAsBoolean()
            && response.has("data") && response.getAsJsonObject("data").has("valid")
            && response.getAsJsonObject("data").get("valid").getAsBoolean()) {
            cache.set("license_status", response.toString());
        }
        return response;
    }

    public JsonObject activateLicense(String licenseKey, String hardwareId) throws ApiException {
        JsonObject data = new JsonObject();
        data.addProperty("action", "activate");
        data.addProperty("license_key", licenseKey);
        data.addProperty("hardware_id", hardwareId);
        JsonObject response = request("license", data);
        if (cache != null) {
            cache.remove("license_status");
        }
        return response;
    }

    public JsonObject deactivateLicense(String licenseKey, String hardwareId) throws ApiException {
        JsonObject data = new JsonObject();
        data.addProperty("action", "deactivate");
        data.addProperty("license_key", licenseKey);
        data.addProperty("hardware_id", hardwareId);
        JsonObject response = request("license", data);
        if (cache != null) {
            cache.remove("license_status");
        }
        return response;
    }

    public JsonObject renewLicense(String licenseKey, Integer extraDays) throws ApiException {
        JsonObject data = new JsonObject();
        data.addProperty("action", "renew");
        data.addProperty("license_key", licenseKey);
        if (extraDays != null) {
            data.addProperty("extra_days", extraDays);
        }
        JsonObject response = request("license", data);
        if (cache != null) {
            cache.remove("license_status");
        }
        return response;
    }

    public JsonObject startTrial(String email, String customerName, String hardwareId, JsonObject customerData)
            throws ApiException {
        JsonObject data = new JsonObject();
        data.addProperty("action", "start");
        data.addProperty("customer_email", email);
        data.addProperty("customer_name", customerName != null ? customerName : "");
        data.addProperty("hardware_id", hardwareId);
        if (customerData != null) {
            for (var entry : customerData.entrySet()) {
                data.add(entry.getKey(), entry.getValue());
            }
        }
        return request("trial", data);
    }

    public JsonObject getTrialStatus(String hardwareId) throws ApiException {
        return request("trial", createPayload("status", "hardware_id", hardwareId));
    }

    public JsonObject convertTrial(String hardwareId, String plan, String name, String email) throws ApiException {
        JsonObject data = new JsonObject();
        data.addProperty("action", "convert");
        data.addProperty("hardware_id", hardwareId);
        if (plan != null) data.addProperty("plan", plan);
        if (name != null && !name.isEmpty()) data.addProperty("customer_name", name);
        if (email != null && !email.isEmpty()) data.addProperty("customer_email", email);
        return request("trial", data);
    }

    public JsonObject replaceDevice(String licenseKey, String oldHardwareId, String newHardwareId)
            throws ApiException {
        JsonObject data = new JsonObject();
        data.addProperty("action", "replace");
        data.addProperty("license_key", licenseKey);
        data.addProperty("old_hardware_id", oldHardwareId);
        data.addProperty("new_hardware_id", newHardwareId);
        JsonObject response = request("device", data);
        if (cache != null) {
            cache.remove("license_status");
        }
        return response;
    }

    public JsonObject bindDevice(String licenseKey, String hardwareId, String deviceName) throws ApiException {
        JsonObject data = new JsonObject();
        data.addProperty("action", "bind");
        data.addProperty("license_key", licenseKey);
        data.addProperty("hardware_id", hardwareId);
        if (deviceName != null && !deviceName.isEmpty()) {
            data.addProperty("device_name", deviceName);
        }
        return request("device", data);
    }

    public JsonObject getProducts() {
        JsonObject payload = new JsonObject();
        payload.addProperty("action", "list");
        if (productId != null && !productId.isEmpty()) {
            payload.addProperty("product_id", productId);
        }
        try {
            return request("store/products", payload);
        } catch (ApiException e) {
            JsonObject fallback = new JsonObject();
            fallback.addProperty("success", false);
            return fallback;
        }
    }

    private static JsonObject createPayload(String action, String key, String value) {
        JsonObject obj = new JsonObject();
        obj.addProperty("action", action);
        obj.addProperty(key, value);
        return obj;
    }
}
